hal/halx86/legacy/halpnpdd.c

```diff
@@ -1,931 +1,544 @@
+                    (PVOID)&Irp->IoStatus.Information);
                 break;
 
-            case IRP_MN_QUERY_BUS_INFORMATION:
-
-                /* Inherit whatever status we had */
-                DPRINT("Query bus information for the PDO\n");
-                Status = Irp->IoStatus.Status;
-                break;
-
-            default:
-
-                /* We don't handle anything else, so inherit the old state */
-                DPRINT1("Illegal IRP: %lx\n", Minor);
-                Status = Irp->IoStatus.Status;
+            case IRP_MN_QUERY_CAPABILITIES:
+                status = HalpQueryCapabilities(
+                    DeviceObject,
+                    ioStack->Parameters.DeviceCapabilities.Capabilities);
                 break;
         }
-
-        /* If it's not supported, inherit the old status */
-        if (Status == STATUS_NOT_SUPPORTED) Status = Irp->IoStatus.Status;
-
-        /* Complete the IRP */
-        DPRINT("IRP completed with status: %lx\n", Status);
-        Irp->IoStatus.Status = Status;
-        IoCompleteRequest(Irp, IO_NO_INCREMENT);
-        return Status;
     }
+
+    Irp->IoStatus.Status = status;
+    IoCompleteRequest(Irp, IO_NO_INCREMENT);
+    return status;
 }
 
+CODE_SEG("PAGE")
 NTSTATUS
 NTAPI
-HalpDispatchWmi(IN PDEVICE_OBJECT DeviceObject,
-                IN PIRP Irp)
+HalpAddDevice(
+    _In_ PDRIVER_OBJECT DriverObject,
+    _In_ PDEVICE_OBJECT TargetDevice)
 {
-    UNIMPLEMENTED_DBGBREAK("HAL: PnP Driver WMI!\n");
+    // Must not be called
+    ASSERT(FALSE);
     return STATUS_SUCCESS;
 }
 
+CODE_SEG("PAGE")
 NTSTATUS
 NTAPI
-HalpDispatchPower(IN PDEVICE_OBJECT DeviceObject,
-                  IN PIRP Irp)
+HalpDispatchPower(
+    _In_ PDEVICE_OBJECT DeviceObject,
+    _In_ PIRP Irp)
 {
-    PFDO_EXTENSION FdoExtension;
+    PAGED_CODE();
+
+    PIO_STACK_LOCATION IrpSp = IoGetCurrentIrpStackLocation(Irp);
+    NTSTATUS Status = Irp->IoStatus.Status;
 
-    DPRINT1("HAL: PnP Driver Power!\n");
-    FdoExtension = DeviceObject->DeviceExtension;
-    if (FdoExtension->ExtensionType == FdoExtensionType)
+    switch (IrpSp->MinorFunction)
     {
-        PoStartNextPowerIrp(Irp);
-        IoSkipCurrentIrpStackLocation(Irp);
-        return PoCallDriver(FdoExtension->AttachedDeviceObject, Irp);
+        case IRP_MN_QUERY_POWER:
+        case IRP_MN_SET_POWER:
+            Status = STATUS_SUCCESS;
+            break;
     }
-    else
+    Irp->IoStatus.Status = Status;
+    PoStartNextPowerIrp(Irp);
+    IoCompleteRequest(Irp, IO_NO_INCREMENT);
+
+    return Status;
+}
+
+CODE_SEG("INIT")
+NTSTATUS
+HalpCreatePdoDeviceObject(
+    _In_ WCHAR PnPID[],
+    _In_ PCM_RESOURCE_LIST BootResources,
+    _In_ UINT32 BootResourcesLength,
+    _In_ PDEVICE_OBJECT Fdo)
+{
+    PFDO_EXTENSION fdoExtension = Fdo->DeviceExtension;
+    PDEVICE_OBJECT devObject;    
+
+    NTSTATUS status = IoCreateDevice(
+        HalpDriverObject,
+        sizeof(PDO_EXTENSION),
```

Comment:
You could declare the `pdoExtension` variable first
```suggestion
        sizeof(*pdoExtension),
```